File: Algorithms/713.Subarray-Product-Less-Than-K/subarray_product_less_than_k.py
# -*- coding: utf-8 -*-

class Solution:
    def numSubarrayProductLessThanK(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        length = len(nums)
        left, right = 0, 0
        res = 0
        prod = 1
        while right < length:
            prod *= nums[right]
            while left <= right and prod >= k:
                prod /= nums[left]
                left += 1
            res += right - left + 1
            right += 1
        return res

File: Algorithms/713.Subarray-Product-Less-Than-K/subarray_product_less_than_k.go
func numSubarrayProductLessThanK(nums []int, k int) int {
	res := 0
	length := len(nums)
	left, right := 0, 0
	prod := 1
	for right = 0; right < length; right++ {
		prod *= nums[right]
		for {
			if left <= right && prod >= k {
				prod = prod / nums[left]
				left += 1
			} else {
				break
			}
		}
		res += right - left + 1
	}
	return res
}
